api_handlers.go

```diff
@@ -0,0 +1,138 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package wptdashboard
+
+import (
+	"encoding/json"
+	"net/http"
+
+	"google.golang.org/appengine"
+	"google.golang.org/appengine/datastore"
+	"net/url"
+)
+
+// apiTestRunsHandler is responsible for emitting test results JSON for all the runs at the given SHA.
+func apiTestRunsHandler(w http.ResponseWriter, r *http.Request) {
+	runSHA, err := GetRunSHA(r)
+	if err != nil {
+		http.Error(w, "Invalid query params", http.StatusBadRequest)
+		return
+	}
+
+	ctx := appengine.NewContext(r)
+	var browserNames []string
+	browserNames, err = GetBrowserNames()
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	var testRuns []TestRun
+	baseQuery := datastore.NewQuery("TestRun").Order("-CreatedAt").Limit(1)
+
+	for _, browserName := range browserNames {
+		var testRunResults []TestRun
+		query := baseQuery.Filter("BrowserName =", browserName)
+		if runSHA != "" && runSHA != "latest" {
+			query = query.Filter("Revision =", runSHA)
+		}
+		if _, err := query.GetAll(ctx, &testRunResults); err != nil {
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+			return
+		}
+		testRuns = append(testRuns, testRunResults...)
+	}
+
+	testRunsBytes, err := json.Marshal(testRuns)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	w.Write(testRunsBytes)
+}
+
+// apiTestRunHandler is responsible for emitting the test result JSON for the given browser's run at the given SHA.
+func apiTestRunHandler(w http.ResponseWriter, r *http.Request) {
+	runSHA, err := GetRunSHA(r)
+	if err != nil {
+		http.Error(w, "Invalid query params", http.StatusBadRequest)
+		return
+	}
+
+	var browserName string
+	browserName, err = getBrowserParam(r)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	} else if browserName == "" {
+		http.Error(w, "Invalid 'browser' param", http.StatusBadRequest)
+		return
+	}
+
+	ctx := appengine.NewContext(r)
+
+	query := datastore.
+		NewQuery("TestRun").
+		Order("-CreatedAt").
+		Limit(1).
+		Filter("BrowserName =", browserName)
+	if runSHA != "" && runSHA != "latest" {
+		query = query.Filter("Revision =", runSHA)
+	}
+
+	var testRuns []TestRun
+	if _, err := query.GetAll(ctx, &testRuns); err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	if len(testRuns) == 0 {
+		http.NotFound(w, r)
+		return
+	}
+
+	testRunsBytes, err := json.Marshal(testRuns[0])
```

Comment:
Can `len(testRuns) > 1` be `true`? Should we have an assertion?

Reply:
If the data manages to get into a 'duplicated-key' state, we gain little by 500'ing the endpoint (rather than just serving the first one we find).

Reply:
I disagree. We should be alerted to the bad data. In particular, if the data have the same `CreatedAt`, then it's feasible for Datastore to deliver us different records each time we hit the endpoint.

Reply:
Right, but it's not an HTTP GET request's job to enforce/sanitize the data layer - if we want to detect or avoid those duplications, we should be doing so as part of the schema, or write phase, or both. Besides which, this is sorted by creation date, so we're surfacing the most recent run (which is intended behaviour, if you somehow re-run for a given SHA)

Reply:
I think the `Limit(1)` will enforce  `testRuns` never being `> 1`, right?

Reply:
Correct